### PR TITLE
Choice of which alerts has to be notified to Slack based on gravity. Slack related fields on preferences disapperar if Slack notifications are OFF 

### PR DESCRIPTION
**Pull request by SauroPollastrini**

### CODE REVIEW

File: include/ntop_defines.h [hunk 1]
```diff
@@ -524,6 +524,7 @@
 #define ALERTS_MANAGER_NOTIFICATION_ENABLED    "ntopng.alerts.notification_enabled"
 #define ALERTS_MANAGER_NOTIFICATION_SENDER     "ntopng.alerts.sender_username"
 #define ALERTS_MANAGER_NOTIFICATION_WEBHOOK    "ntopng.alerts.slack_webhook"
+#define ALERTS_MANAGER_NOTIFICATION_GRAVITY    "ntopng.prefs.slack_alert_gravity"
```
Comment (simonemainardi) [inline include/ntop_defines.h]: gravity is awkward has another meaning.

use word 'level' or 'severity' that are the standard terms ntopng uses.

File: src/AlertsManager.cpp [hunk 2]
```diff
@@ -687,10 +703,20 @@ void AlertsManager::notifyAlert(AlertEntity alert_entity, const char *alert_enti
...
 			 json_object_new_string(getAlertLevel(alert_severity)));
 
+
   if(ntop->getRedis()->get((char*)ALERTS_MANAGER_SENDER_USERNAME,
-			   alert_sender_name, sizeof(alert_sender_name)) >= 0)
+			   alert_sender_name, sizeof(alert_sender_name)) >= 0) {
+
+    strncpy(notification_username, alert_sender_name, 64);
+
+    if (alert_severity == alert_level_error) strncat(notification_username, " - NOTIFY OF ERROR", 32);
```
Comment (simonemainardi) [inline src/AlertsManager.cpp:712]: avoid hard-coding strings here. Also see if you can re-use this: https://github.com/SauroPollastrini/ntopng/blob/dev/src/AlertsManager.cpp#L649

File: src/AlertsManager.cpp [hunk 3]
```diff
@@ -713,21 +739,59 @@ void AlertsManager::notifyAlert(AlertEntity alert_entity, const char *alert_enti
...
+
+  char buf[4], choice[32];
+  bool alert_to_be_notified=false; //indicates if the notification has to be done according to the slack notifications preference
+  SlackNotificationChoice notification_choice;
+
+  ntop->getTrace()->traceEvent(TRACE_NORMAL, 
+    "Entrata alert %s %s %s %s", getAlertLevel(alert_severity),
+    alert_entity_value, alert_origin, alert_target );
+
+  if((ntop->getRedis()->get((char*)ALERTS_MANAGER_NOTIFICATION_ENABLED,
```
Comment (simonemainardi) [inline src/AlertsManager.cpp:755]: avoid Redis reads every time. Cache the values in a global prefs variable and use it. See this: https://github.com/SauroPollastrini/ntopng/blob/dev/include/Prefs.h#L154

make sure to refresh the variable when the user submit changes from the ntopng preferences UI.

File: src/AlertsManager.cpp [hunk 4]
```diff
@@ -713,21 +739,59 @@ void AlertsManager::notifyAlert(AlertEntity alert_entity, const char *alert_enti
...
+          buf, sizeof(buf)) >= 0)
+     && (!strcmp(buf, "1"))
+    ) 
+
+  {
+
+    ntop->getRedis()->get((char*) ALERTS_MANAGER_NOTIFICATION_GRAVITY, choice, sizeof(choice));
```
Comment (simonemainardi) [inline src/AlertsManager.cpp]: see above

File: src/AlertsManager.cpp [hunk 5]
```diff
@@ -713,21 +739,59 @@ void AlertsManager::notifyAlert(AlertEntity alert_entity, const char *alert_enti
 
 /* **************************************************** */
 
+void AlertsManager::notifyToSlackIfNeeded (AlertEntity alert_entity, const char *alert_entity_value,
```
Comment (simonemainardi) [inline src/AlertsManager.cpp:742]: notifyAlert and notifyToSlackIfNeeded do similar things and it is confusing to know when to call one or the other. Please refactor to make it more clear

File: src/AlertsManager.cpp [hunk 6]
```diff
@@ -671,13 +678,22 @@ const char* AlertsManager::getAlertType(AlertType alert_type) {
...
+
+/* **************************************************** */
 
 void AlertsManager::notifyAlert(AlertEntity alert_entity, const char *alert_entity_value,
 				AlertType alert_type, AlertLevel alert_severity,
 				const char *alert_json,
 				const char *alert_origin, const char *alert_target) {
   json_object *notification;
-  char alert_sender_name[64], message[2015];
+  char alert_sender_name[64], message[2015], notification_username[96];
```
Comment (simonemainardi) [inline src/AlertsManager.cpp:696]: don't hardcode sizes here. Look in ntop_defines and look for already defined constants such as https://github.com/SauroPollastrini/ntopng/blob/dev/include/ntop_defines.h#L277

File: third-party/json-c/config.h.in [hunk 7]
```diff
@@ -104,8 +104,7 @@
 /* Public define for json_inttypes.h */
 #undef JSON_C_HAVE_INTTYPES_H
 
-/* Define to the sub-directory in which libtool stores uninstalled libraries.
-   */
+/* Define to the sub-directory where libtool stores uninstalled libraries. */
```
Comment (simonemainardi) [inline third-party/json-c/config.h.in]: this file shouldn't be part of the commit

**Review by simonemainardi — CHANGES_REQUESTED** [review]

File: include/ntop_defines.h (hunk 1 above)
Comment (SauroPollastrini) [reply include/ntop_defines.h]: Changed in 'severity'

File: third-party/json-c/config.h.in (hunk 7 above)
Comment (SauroPollastrini) [reply third-party/json-c/config.h.in]: Removed

File: src/AlertsManager.cpp (hunk 3 above)
Comment (lucaderi) [reply src/AlertsManager.cpp:755]: Remove non-English messages

**Comment by lucaderi** [comment]
I will merge the code and handle the fixes myself to expedite this merge
